File: frontend/src/models/GeoTypes.ts
/**
 * GeoTypes - TypeScript interfaces matching backend DTOs
 * Defines data structures for geographic coordinates, polygons, and API requests/responses
 */

/**
 * Represents a geographic coordinate (latitude, longitude)
 */
export interface GeoPoint {
  latitude: number // WGS84, degrees [-90, 90]
  longitude: number // WGS84, degrees [-180, 180]
}

/**
 * Represents a closed polygon boundary
 */
export interface Polygon {
  id: string // UUID
  vertices: GeoPoint[] // ≥ 3 points for valid polygon
  isExcludePolygon: boolean // true if this is an exclude area
  parentPolygonId?: string // set if this is an exclude polygon
  isValid: boolean // computed: vertices.length >= 3
  hasIntersections: boolean // computed: sides cross each other
  computedAreaSquareMeters: number // geodetic area
  computedPerimeterMeters: number // geodetic perimeter
  perSideLengthsMeters: number[] // array of distances between consecutive vertices
}

/**
 * Represents a distance measurement as a polyline
 */
export interface Measurement {
  id: string // UUID
  vertices: GeoPoint[] // ≥ 2 points for valid measurement
  isValid: boolean // computed: vertices.length >= 2
  totalDistanceMeters: number // total distance along polyline
  perSegmentDistancesMeters: number[] // array of segment distances
}

/**
 * User's unit preferences
 */
export interface UnitPreference {
  areaUnit: 'acres' | 'hectares' | 'sqm'
  distanceUnit: 'feet' | 'meters' | 'miles' | 'km'
}

/**
 * Session state
 */
export interface SessionState {
  primaryPolygon?: Polygon
  excludePolygons: Polygon[]
  unitPreference: UnitPreference
}

// ============================================================================
// API Request/Response Types
// ============================================================================

/**
 * POST /geometry/calculate-area request
 */
export interface CalculateAreaRequest {
  vertices: GeoPoint[] // primary polygon vertices
  excludePolygons?: GeoPoint[][] // optional exclude polygons
}

/**
 * POST /geometry/calculate-area response
 */
export interface CalculateAreaResponse {
  areaSquareMeters: number
  netAreaSquareMeters: number
  excludedAreaSquareMeters: number
  perSideLengthsMeters: number[]
  hasIntersections: boolean
}

/**
 * POST /geocode/search request
 */
export interface GeocodeSearchRequest {
  query: string // e.g., "123 Main St, Springfield, IL"
  maxResults?: number // default 5
}

/**
 * Search result from geocoding
 */
export interface GeocodeResult {
  id: string
  displayName: string
  latitude: number
  longitude: number
  boundingBox?: {
    minLat: number
    maxLat: number
    minLon: number
    maxLon: number
  }
}

/**
 * POST /geocode/search response
 */
export interface GeocodeSearchResponse {
  results: GeocodeResult[]
}

/**
 * POST /geocode/reverse request
 */
export interface ReverseGeocodeRequest {
  latitude: number
  longitude: number
}

/**
 * POST /geocode/reverse response
 */
export interface ReverseGeocodeResponse {
  address: string
  latitude: number
  longitude: number
}

File: backend/src/SteelTree.GeoAcre.Web.Api/Models/GeoTypes.cs
namespace SteelTree.GeoAcre.Web.Api.Models;

/// <summary>
/// Request DTO for geometry area calculation.
/// </summary>
public class CalculateAreaRequest
{
    /// <summary>
    /// Primary polygon vertices in order (latitude, longitude).
    /// </summary>
    public required LatLngDto[] Vertices { get; set; }

    /// <summary>
    /// Optional list of exclude polygons to subtract from the primary polygon.
    /// </summary>
    public LatLngDto[][]? ExcludePolygons { get; set; }
}

/// <summary>
/// Response DTO for geometry area calculation.
/// </summary>
public class CalculateAreaResponse
{
    /// <summary>
    /// Area of the primary polygon in square meters.
    /// </summary>
    public required double AreaSquareMeters { get; set; }

    /// <summary>
    /// Net area after subtracting excludes in square meters.
    /// </summary>
    public double NetAreaSquareMeters { get; set; }

    /// <summary>
    /// Total area of all exclude polygons in square meters.
    /// </summary>
    public double ExcludedAreaSquareMeters { get; set; }

    /// <summary>
    /// Length of each side of the primary polygon in meters.
    /// </summary>
    public required double[] PerSideLengthsMeters { get; set; }

    /// <summary>
    /// Whether the primary polygon has self-intersections.
    /// </summary>
    public required bool HasIntersections { get; set; }
}

/// <summary>
/// Request DTO for geocoding search.
/// </summary>
public class GeocodeSearchRequest
{
    /// <summary>
    /// Search query (address or place name).
    /// </summary>
    public required string Query { get; set; }

    /// <summary>
    /// Maximum number of results to return.
    /// </summary>
    public int MaxResults { get; set; } = 10;
}

/// <summary>
/// Response DTO for geocoding search result.
/// </summary>
public class GeocodeSearchResponse
{
    /// <summary>
    /// List of search results.
    /// </summary>
    public required GeocodeResult[] Results { get; set; }
}

/// <summary>
/// Individual geocoding result.
/// </summary>
public class GeocodeResult
{
    /// <summary>
    /// Unique identifier for this result.
    /// </summary>
    public required string Id { get; set; }

    /// <summary>
    /// Display name of the location.
    /// </summary>
    public required string DisplayName { get; set; }

    /// <summary>
    /// Latitude of the location.
    /// </summary>
    public required double Latitude { get; set; }

    /// <summary>
    /// Longitude of the location.
    /// </summary>
    public required double Longitude { get; set; }

    /// <summary>
    /// Bounding box of the location (min_lat, max_lat, min_lon, max_lon).
    /// </summary>
    public double[]? BoundingBox { get; set; }
}

/// <summary>
/// Request DTO for reverse geocoding.
/// </summary>
public class ReverseGeocodeRequest
{
    /// <summary>
    /// Latitude of the location.
    /// </summary>
    public required double Latitude { get; set; }

    /// <summary>
    /// Longitude of the location.
    /// </summary>
    public required double Longitude { get; set; }
}

/// <summary>
/// Response DTO for reverse geocoding.
/// </summary>
public class ReverseGeocodeResponse
{
    /// <summary>
    /// Formatted address string.
    /// </summary>
    public required string Address { get; set; }

    /// <summary>
    /// Latitude of the location.
    /// </summary>
    public required double Latitude { get; set; }

    /// <summary>
    /// Longitude of the location.
    /// </summary>
    public required double Longitude { get; set; }
}

/// <summary>
/// Latitude/Longitude coordinate pair.
/// </summary>
public class LatLngDto
{
    /// <summary>
    /// Latitude (-90 to 90).
    /// </summary>
    public required double Latitude { get; set; }

    /// <summary>
    /// Longitude (-180 to 180).
    /// </summary>
    public required double Longitude { get; set; }
}
